batch/batch/server/server.py

```diff
@@ -117,11 +162,61 @@ def _read_log(self):
                     pass
             return None
         if self._state == 'Complete':
-            return _read_file(_log_path(self.id))
+            return _read_file(_log_path(self.job_id, self.name))
         assert self._state == 'Cancelled'
         return None
 
-    def __init__(self, pod_spec, batch_id, attributes, callback, parent_ids, scratch_folder):
+    def mark_unscheduled(self):
+        if self._pod_name:
+            del pod_name_job_task[self._pod_name]
+            self._pod_name = None
+        self.create_pod()
+
+    def mark_complete(self, pod):
+        self.exit_code = pod.status.container_statuses[0].state.terminated.exit_code
+
+        pod_log = v1.read_namespaced_pod_log(
+            pod.metadata.name,
+            POD_NAMESPACE,
+            _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+
+        add_event({'message': f'job {self.job_id}, {self.name} task exited', 'log': pod_log[:64000]})
+
+        fname = _log_path(self.job_id, self.name)
+        with open(fname, 'w') as f:
+            f.write(pod_log)
+        log.info(f'wrote log for job {self.job_id}, {self.name} task to {fname}')
+
+        if self._pod_name:
+            del pod_name_job_task[self._pod_name]
+            self._pod_name = None
+
+        self.set_state('Complete')
+        self.job.mark_complete()
+
+
+class Job:
+    def _next_task(self):
```

Comment:
this is a bit confusing since it mutates _current_task and returns it, can we change the name or make it not mutate?

Reply:
Doesn't calling next on an iterator do the same thing? Doesn't increment the counter and return the next value?